File: implement-shell-tools/wc/wc.py
import os
import sys

# --- Step 1: Parse command-line arguments ---
# Separate flags (-l, -w, -c) from file paths.
args = sys.argv[1:]

count_lines = False  # -l flag
count_words = False  # -w flag
count_bytes = False  # -c flag
files = []

for arg in args:
    if arg == "-l":
        count_lines = True
    elif arg == "-w":
        count_words = True
    elif arg == "-c":
        count_bytes = True
    else:
        files.append(arg)

# If no flags are given, show all three counts (lines, words, bytes)
no_flags = not count_lines and not count_words and not count_bytes

# --- Step 2: Count lines, words, and bytes for each file ---
# Store the results so we can calculate totals for multiple files.
results = []

for file in files:
    with open(file, "r", encoding="utf-8") as f:
        content = f.read()

    # Byte count comes from the file size on disk, not the string length.
    # (Important for multi-byte UTF-8 characters where bytes > chars.)
    byte_count = os.path.getsize(file)

    # Lines = number of newline characters.
    line_count = content.count("\n")

    # Words: Python's str.split() with no argument splits on ANY run of whitespace
    # AND drops empty strings — no separate filter step needed (unlike JS).
    word_count = len(content.split())

    results.append({
        "file": file,
        "lines": line_count,
        "words": word_count,
        "bytes": byte_count,
    })

# --- Step 3: Format and print output ---
# Each number is right-justified in an 8-character field, followed by a space and filename.

def format_row(lines, words, bytes_, name):
    row = ""

    if no_flags:
        # No flags: show all three counts
        row += str(lines).rjust(8)
        row += str(words).rjust(8)
        row += str(bytes_).rjust(8)
    else:
        # Only show the requested counts, in order: lines, words, bytes
        if count_lines:
            row += str(lines).rjust(8)
        if count_words:
            row += str(words).rjust(8)
        if count_bytes:
            row += str(bytes_).rjust(8)

    row += f" {name}"
    return row


for r in results:
    print(format_row(r["lines"], r["words"], r["bytes"], r["file"]))

# If there are multiple files, print a "total" row
if len(results) > 1:
    total_lines = sum(r["lines"] for r in results)
    total_words = sum(r["words"] for r in results)
    total_bytes = sum(r["bytes"] for r in results)
    print(format_row(total_lines, total_words, total_bytes, "total"))

File: implement-shell-tools/wc/wc.js
const fs = require("fs");

// --- Step 1: Parse command-line arguments ---
// Separate flags (-l, -w, -c) from file paths.
const args = process.argv.slice(2);

let countLines = false; // -l flag
let countWords = false; // -w flag
let countBytes = false; // -c flag
const files = [];

for (const arg of args) {
  if (arg === "-l") {
    countLines = true;
  } else if (arg === "-w") {
    countWords = true;
  } else if (arg === "-c") {
    countBytes = true;
  } else {
    files.push(arg);
  }
}

// If no flags are given, show all three counts (lines, words, bytes)
const noFlags = !countLines && !countWords && !countBytes;

// --- Step 2: Count lines, words, and bytes for each file ---
// We store the results so we can calculate totals for multiple files.
const results = [];

for (const file of files) {
  const content = fs.readFileSync(file, "utf-8");
  const bytes = fs.statSync(file).size; // byte count from file size (not string length)
  const lines = content.split("\n").length - 1; // number of newline characters
  // Words: split on whitespace, filter out empty strings from leading/trailing space
  const words = content.split(/\s+/).filter((w) => w.length > 0).length;

  results.push({ file, lines, words, bytes });
}

// --- Step 3: Format and print output ---
// Each number is right-justified in an 8-character field, followed by a space and filename.

function formatRow(lines, words, bytes, name) {
  let row = "";

  if (noFlags) {
    // No flags: show all three counts
    row += String(lines).padStart(8, " ");
    row += String(words).padStart(8, " ");
    row += String(bytes).padStart(8, " ");
  } else {
    // Only show the requested counts, in order: lines, words, bytes
    if (countLines) row += String(lines).padStart(8, " ");
    if (countWords) row += String(words).padStart(8, " ");
    if (countBytes) row += String(bytes).padStart(8, " ");
  }

  row += ` ${name}`;
  return row;
}

for (const r of results) {
  console.log(formatRow(r.lines, r.words, r.bytes, r.file));
}

// If there are multiple files, print a "total" row
if (results.length > 1) {
  const totalLines = results.reduce((sum, r) => sum + r.lines, 0);
  const totalWords = results.reduce((sum, r) => sum + r.words, 0);
  const totalBytes = results.reduce((sum, r) => sum + r.bytes, 0);
  console.log(formatRow(totalLines, totalWords, totalBytes, "total"));
}
